compiler/rustc_resolve/src/late.rs

```diff
@@ -1331,6 +1343,10 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
         opt_ns: Option<Namespace>, // `None` indicates a module path in import
         finalize: Option<Finalize>,
     ) -> PathResult<'a> {
+        if self.in_delegation_item && path.len() == 1 && path[0].ident.name == kw::SelfLower {
+            return PathResult::NonModule(def::PartialRes::new(def::Res::Err));
```

Comment:
This is hacky and subtly incorrect.

I would suggest creating a new `Res::Local(id)` binding for `self` when entering the delegation item body.
Then all its uses in the body will automatically be resolved correctly.
`fresh_binding` can be used for that (see its other uses for examples).
`in_delegation_item` will no longer be necessary in that case.

Node id can be taken from the delegation item.
The a pass similar to `SelfResolver` will be required to update it to the first parameter's id.
Or you can reserve a `NodeId` for the future first parameter in advance.
Not sure what would be simpler (there may be issues if the id is reserved, but the first parameter won't actually exist).